### PR TITLE
Add support for Poisson, Exponential, and Bernoulli likelihoods w/ arbitrary link functions

### LINKED ISSUE #3: Add support for arbitrary link functions
`blm` currently supports only the identity link function for linear response variables. We should also allow the user to easily specify alternative link functions to make it a true Bayesian GLM.

There are two options here:

1. Provide pre-built implementations of common link functions and their respective distributions.
2. Allow the user to specify whatever link they want via a function F: yhat -> D, where yhat is the linear predictor and D is the response distribution.

Option 2 seems nicer but has the technical difficulty of allowing for arbitrary priors and variable names which depend on the link function and are not known ahead of time.

### PR DESCRIPTION
**Pull request by bgroenks96**
This PR accomplishes three things:

1. Adds support for a few non-normal likelihoods w/ arbitrary link functions (canonical link functions are provided by default). The appropriate likelihood is automatically selected based on the response (LHS) term type and the data type, i.e. (`ContinuousTerm`+`<:Real` => `Normal`, `ContinuousTerm`+`<:Integer` => `Poisson`, `CategoricalTerm` => `Bernoulli`). The latter will need to change a bit if/when we add support for categorical or multinomial likelihoods.
2. Refactors the code a bit to accommodate (1). The most notable change is that `bayesreg` is now the abstract Bayesian regression function and `blm` invokes `bayesreg` using a `lm` variant as the Turing model. Priors are now also handled a bit more elegantly (i.e. wrapped up into `kwargs` and then named tuples for unpacking with `Parameters.jl`.
3. Adds more examples to the README as well as tests to go along with them.

Resolves #3

### CODE REVIEW

**Comment by bgroenks96** [comment]
One possible "improvement" to be made here would be to rearrange the parameters a bit and try to maintain some semblance of parity with `GLM.jl`, i.e.

```julia
blm(formula, data, MCMC(NUTS(),1000), Bernoulli(), LogitLink(),...)
```
I haven't really used the `GLM` package personally, so I didn't really think too much about it, but in hindsight, might be a good idea. On the other hand, I kind of like the elegance of not needing separate types for the link functions in the way I currently have it.

File: src/blm.jl [hunk 1]
```diff
@@ -40,48 +10,127 @@ function postprocess(::MCMC, chain::MCMCChains.Chains, f::FormulaTerm, df::DataF
...
-    blm(f::FormulaTerm, df::DataFrame, inferencestrat::InferenceStrategy=MCMC(NUTS(),1000); alphaprior, betaprior, sigmaprior)
+Internal function that specifies default priors for non-Normal likelihood BLMs. Additional priors for custom model
+definitions may be passed in `kwargs`.
+"""
+blmpriors(
+    ::Type{<:Distribution};
+    α_prior = Normal(),
+    β_prior = (X,y) -> MvNormal(zeros(size(X,2)),1),
+    kwargs...
+) = (α_prior=α_prior,β_prior=β_prior,kwargs...)
```
Comment (just-cameron) [inline src/blm.jl]: Multiline stuff should be in `function ... end` syntax, usually -- minor thing, but I think it tends to make things more readable.

```suggestion
function blmpriors(
    ::Type{<:Distribution};
    α_prior = Normal(),
    β_prior = (X,y) -> MvNormal(zeros(size(X,2)),1),
    kwargs...
)
    return (α_prior=α_prior,β_prior=β_prior,kwargs...)
end
```

File: src/blm.jl [hunk 2]
```diff
@@ -40,48 +10,127 @@ function postprocess(::MCMC, chain::MCMCChains.Chains, f::FormulaTerm, df::DataF
...
+"""
+Internal function that specifies default priors for Normal likelihood BLMs.
+"""
+blmpriors(
+    ::Type{Normal};
+    α_prior = Normal(),
+    β_prior = (X,y) -> MvNormal(zeros(size(X,2)),1),
+    σ_prior = InverseGamma(2,3),
+    kwargs...
+) = (α_prior=α_prior,β_prior=β_prior,σ_prior=σ_prior,kwargs...)
```
Comment (just-cameron) [inline src/blm.jl]:
```suggestion
function blmpriors(
    ::Type{Normal};
    α_prior = Normal(),
    β_prior = (X,y) -> MvNormal(zeros(size(X,2)),1),
    σ_prior = InverseGamma(2,3),
    kwargs...
)
    return (α_prior=α_prior,β_prior=β_prior,σ_prior=σ_prior,kwargs...)
end
```

**Review by just-cameron — COMMENTED** [review]
Minor style notes, but other than that, I am amped for this one. Very slick stuff.

**Comment by just-cameron** [comment]
I'm just going to commit and merge the minor style stuff I added -- thanks!

**Pull request merged**